### ISSUE TITLE
Azure Source Enhancement: Consumption Data

### ISSUE THREAD

**Issue by tomtom215** [cloudquery/cloudquery#8929]
### Is your feature request related to a problem? Please describe.

We are testing CloudQuery in our Azure Enterprise subscription and would like to get more consumption data than just the cost report data that the current source plugin provides.

### Describe the solution you'd like

I would like to request the enhancement of the Azure Source plugin to include the possibility to include Azure Enterprise Consumption API data found here https://learn.microsoft.com/en-us/rest/api/consumption/

### Describe alternatives you've considered

Running CloudQuery Azure Source Plugin then manually running a batch of API requests to the Consumption API to enhance the CloudQuery data
### Additional context

I understand that this would require an Azure Enterprise subscription to test, so I'd be happy to try out any PR's

**Comment by tomtom215** [comment]
Thanks for such a quick turn around @erezrokah ! We will be able to to test this early next week (this coming Monday/Tuesday) and will report any findings here. Please let us know if there's anything specific you'd like us to look into or try out

**Comment by erezrokah** [comment]
Hi @tomtom215, thanks for the quick response! I marked https://github.com/cloudquery/cloudquery/pull/9117 as ready for review and followed up with a few updates based on my tests. Mostly I moved some APIs to use a billing profile scope as it that should work on all accounts, and also required to get the "Modern" responses.

As we don't have an enterprise account I couldn't validate all the data, but I did validate the API calls don't return an error (some return 404 or 204, but it appears that's expected based on the content of the error I got).

One thing to pay attention to is the `azure_consumption_billing_profile_reservation_recommendations`, `azure_consumption_billing_account_usage_details` and `azure_consumption_billing_account_charges` tables as those cast the response from an interface to a struct, see:
https://github.com/erezrokah/cloudquery/blob/c2664ed23a899733b2032ba04cd4b659586a76b7/plugins/source/azure/resources/services/consumption/billing_profile_reservation_recommendations.go#L43
https://github.com/erezrokah/cloudquery/blob/c2664ed23a899733b2032ba04cd4b659586a76b7/plugins/source/azure/resources/services/consumption/billing_account_usage_details.go#L35
https://github.com/erezrokah/cloudquery/blob/c2664ed23a899733b2032ba04cd4b659586a76b7/plugins/source/azure/resources/services/consumption/billing_account_charges.go#L33

The code is based on the docs, but I can't verify it without running on real resources

**Comment by ethankelly99** [comment]
Hi @erezrokah! 
@tomtom215 and I got a chance to test this today. I'll try summarise our findings below.

Unfortunately we only have "Legacy" Billing Accounts. The current implementation only supports "Modern" Billing Accounts. Would you consider adding support for Legacy billing accounts?

A Legacy billing account has a simple GUID identifier e.g. `9971ccb0-02bb-45e2-bd6a-9e340372dcba` whereas a Modern billing account has a multi-part identifier that seems to separated with a colon e.g. `7c05a543-80ff-571e-9f98-1063b3b53cf2:99ad03ad-2d1b-4889-a452-090ad407d25f_2019-05-31` (I pulled these billing account IDs from a blog post, so they're not private and don't belong to us 😄). I haven't looked into this in too much detail yet, but I think when discovering the billing accounts, the API returns a field that indicates if an account is legacy or modern. 

Here's an example of the response when a Legacy billing account ID is queried using the Modern API:
```
GET https://management.azure.com/providers/Microsoft.Billing/billingAccounts/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/providers/Microsoft.Consumption/usageDetails
--------------------------------------------------------------------------------
RESPONSE 400: 400 Bad Request
ERROR CODE: 400
--------------------------------------------------------------------------------
{
  "error": {
    "code": "400",
    "message": "The request could not be understood. Parameter BillingAccountId: xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx is in a format that could not be processed with this version (Request ID: xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx)"
  }
}
--------------------------------------------------------------------------------
```

There's a few tables we haven't been able to test yet as our users are missing the required permissions. We are currently waiting for them to be granted internally and once this is done, we will re-run the tests. But I'd imagine they'll face the same issue with the Legacy vs. Modern billing accounts.
- `azure_consumption_billing_account_budgets`
- `azure_consumption_billing_account_reservation_details`
- `azure_consumption_billing_account_reservation_recommendations`
- `azure_consumption_billing_account_reservation_summaries`
- `azure_consumption_billing_account_reservation_transactions`

We're happy to test this further, but unfortunately won't be much help with Modern billing accounts as we just don't have any :)

**Comment by erezrokah** [comment]
Hi @ethankelly99 👋 Thanks for testing 🚀 Yes I think we can add both legacy and modern. Those would have to be in different tables as the API returns different structs for legacy and modern, so we would have `azure_consumption_billing_account_legacy_usage_details` and `azure_consumption_billing_account_modern_usage_details` for example. Would that work for you?

**Comment by ethankelly99** [comment]
@erezrokah, my pleasure, thanks for putting the PR together! Yes, I think that would work for us 👍

**Comment by erezrokah** [comment]
> I haven't looked into this in too much detail yet, but I think when discovering the billing accounts, the API returns a field that indicates if an account is legacy or modern.

Can you point me to the field that says the account type (legacy or modern)? Looking at https://learn.microsoft.com/en-us/rest/api/billing/2019-10-01-preview/billing-accounts/list?tabs=HTTP#billingaccountslistwithexpand I couldn't find it.
I don't mind deducing it from the ID, but rather use a dedicate field.
Maybe you can share a redacted version of the `azure_billing_accounts` and `azure_billing_enrollment_accounts` tabes after a successful sync for your organization so I can compare with the data I see?

> Here's an example of the response when a Legacy billing account ID is queried using the Modern API:

OK, so I'm a bit confused by this as it doesn't match what the docs say (which means those are probably wrong).
A request to legacy usage details should look the same as a request to modern, just with a different ID, see:
https://learn.microsoft.com/en-us/rest/api/consumption/usage-details/list?tabs=HTTP#billingaccountusagedetailslist-legacy
```
GET https://management.azure.com/providers/Microsoft.Billing/BillingAccounts/1234/providers/Microsoft.Consumption/usageDetails?api-version=2021-10-01
```

https://learn.microsoft.com/en-us/rest/api/consumption/usage-details/list?tabs=HTTP#billingaccountusagedetailslist-modern
```
GET https://management.azure.com/providers/Microsoft.Billing/BillingAccounts/1234:56789/providers/Microsoft.Consumption/usageDetails?api-version=2021-10-01
```

Did you get that error when using CloudQuery? I believe the API version is missing from that request (maybe it just not logged with the error)

**Comment by erezrokah** [comment]
I added another [commit](https://github.com/cloudquery/cloudquery/pull/9117/commits/d3da081e8c6c8cfc09c5b4b9ce9682a686a2752a) to the PR to split usage details into separate tables (based on the ID format), and will do so for other "legacy" resources based on the docs.

I'm still not sure it as I'm using the same API for legacy and modern usage details, only passing a different account ID format.
If that doesn't work I'll probably reach out to Azure support, and we can also try one of the other "legacy" APIs here https://learn.microsoft.com/en-us/rest/api/consumption/usage-details/list?tabs=HTTP#examples

**Comment by ethankelly99** [comment]
> Can you point me to the field that says the account type (legacy or modern)?

I double checked the docs and also tested the API locally - unfortunately I was wrong, the API doesn't return the account type. Sorry for the confusion! I must've been thinking of a different endpoint.

> OK, so I'm a bit confused by this as it doesn't match what the docs say

I need to revise some of the feedback from my last comment. It turns out your last comment (and the API docs) are correct, requests for usage details for legacy and modern accounts are the same endpoint, just with different IDs. I made a mistake when testing by using my user which I thought had Billing Reader access to our Enterprise billing accounts, but it only has access to my individual billing account (which is still on our corporate tenant, but is an "Individual" billing account instead).

So the reason for the error was that CloudQuery discovered my individual billing account and tried to query the enterprise-only consumption APIs with it.

Sorry for causing wasted time 🤦 I'm still waiting for the required permissions to be granted to my user (and a service principal that we're using) to access the enterprise billing accounts (we're hoping it'll be granted this week). Once I have this, I'll test from scratch again.

> Did you get that error when using CloudQuery? I believe the API version is missing from that request (maybe it just not logged with the error)

Yep I got that error when using CloudQuery (I just copied it out and formatted it so it was easier to read), it seems the API version isn't included in the error.

**Comment by ethankelly99** [comment]
Hi @erezrokah!
I'm finally able to follow up on this, sorry it's taken so long! Unfortunately it turns out it isn't possible for us to get permissions on the enterprise billing accounts as it would include data from subscriptions that are out of scope of our investigation. Would you consider supporting the subscription scope as well? This is something we already have access to and would be happy to test 😄

**Comment by erezrokah** [comment]
> Would you consider supporting the subscription scope as well? This is something we already have access to and would be happy to test 

Yes I think we could do that. FYI not all APIs support a subscription scope, e.g. https://learn.microsoft.com/en-us/rest/api/consumption/charges/list?tabs=HTTP#uri-parameters

**Comment by erezrokah** [comment]
So, so I've added the subscription based tables I could find, see https://github.com/cloudquery/cloudquery/pull/9117#issuecomment-1470158742

Please let me know if there's something I missed.

**Comment by ethankelly99** [comment]
Awesome, thank you Erez!

The following Consumption APIs also support the subscription scope:
- [Usage Details](https://learn.microsoft.com/en-us/rest/api/consumption/usage-details/list?tabs=HTTP)
- [Price Sheet](https://learn.microsoft.com/en-us/rest/api/consumption/price-sheet/get?tabs=HTTP)
- [Tags](https://learn.microsoft.com/en-us/rest/api/consumption/tags/get?tabs=HTTP)

I'm kinda surprised that some don't list support for the subscription scope in the docs. Out of curiosity I will just try some subscription-scoped requests locally, e.g. [Reservations Details](https://learn.microsoft.com/en-us/rest/api/consumption/reservations-details/list?tabs=HTTP)

**Comment by erezrokah** [comment]
Thanks for sharing those, I missed tags completely so will add those for all relevant scopes.

For Price Sheet is seems the API returns a single item? I think that fine, just confirming.
For Usage Details, do you know if subscription scope returns "legacy" or "modern" type of response? If not I'll assume legacy as that's the case for Reservation Recommendation

**Comment by erezrokah** [comment]
OK so I've updated the PR based on your useful examples, see [same comment](https://github.com/cloudquery/cloudquery/pull/9117#issuecomment-1470158742).

A few notes:
- I couldn't test these as it seems our subscription does not support them
- For Price Sheets I'm using the [API without a billing period parameter](https://learn.microsoft.com/en-us/rest/api/consumption/price-sheet/get?tabs=HTTP). To list by billing periods I think we should need to use the information from [`azure_billing_periods`.](https://www.cloudquery.io/docs/plugins/sources/azure/tables/azure_billing_periods), assuming it has what we need.

**Comment by erezrokah** [comment]
> Hey @erezrokah!
> This is awesome! I tested the updated PR just now. It worked without any errors :)

This is great 🚀 

> Do you think it would be achievable to add this?

Yes, we could do this. We could get the periods during discovery phase like we do with accounts and use that information later on (probably for other tables too, and not just `azure_consumption_subscription_legacy_usage_details`).


> But a potential problem with the current implementation is that the [properties attribute of the PriceSheetResult](https://learn.microsoft.com/en-us/rest/api/consumption/price-sheet/get?tabs=HTTP#pricesheetresult) response is paginated and only the first page is fetched/stored.

Ah, good catch I missed it. Let me look how we could paginate over the result of a `Get` call. We only paginate on `List` call, so I'll need to check what the Azure SDK exposes for doing it for `Get`.

**Comment by erezrokah** [comment]
Hi @ethankelly99, sorry for the delay, we're in the middle of migrating to a new type system, so just had the time to get back to this now. I updated the PR based on your comment. See https://github.com/cloudquery/cloudquery/pull/9117#issuecomment-1516606448

Based on the docs in order to get usage details for a billing period, we need to add it to the scope, see:
https://learn.microsoft.com/en-us/rest/api/consumption/usage-details/list?tabs=HTTP#uri-parameters
![image](https://user-images.githubusercontent.com/26760571/233427031-76c3ae48-bdee-4c65-845b-b4c665b683df.png)

so I did that instead of the filter parameter. Can you verify this works for you?

For the pagination over prices sheets it's a bit hacky at the moment, and I'm waiting on a response from https://github.com/Azure/azure-sdk-for-go/issues/20671

Finally it seems Azure added a note that the usage details API is going to be deprecated (not sure when).
I recommend looking at https://www.cloudquery.io/blog/azure-cost-optimization-with-cloudquery that uses the Cost API instead to see if it matches your needs.

**Comment by erezrokah** [comment]
Hi 👋 This issue was auto-closed as we merged the consumption PR. It's released in [`v7.2.0`](https://github.com/cloudquery/cloudquery/releases/tag/plugins-source-azure-v7.2.0) of the Azure source plugin, as we wanted to push these new tables to more users and see if any other issues come up.

We might still need to do a bit of tuning to the code, which we're happy to do in follow up PRs.
Please comment if the issue is not resolved and we can re-open it or open a new one with your requests.
